### PR TITLE
Sparsehistogram: SHS chunk recording and head cutting to head block

### PR DESCRIPTION
**Pull request by Dieterbe**
experimental sparsehistogram stuff for @beorn7 and @codesome

### CODE REVIEW

File: tsdb/chunkenc/histo.go [hunk 1]
```diff
@@ -387,14 +386,16 @@ func (a *histoAppender) recode(posInterjections, negInterjections []interjection
...
 	// adopt the new appender into ourselves
 	// we skip porting some fields like schema, t, cnt and zcnt, sum because they didn't change between our old chunk and the recoded one
-	app2 := app.(*histoAppender)
+	app2 := app.(*HistoAppender)
 	a.b = app2.b
 	a.posSpans, a.negSpans = posSpans, negSpans
 	a.posbuckets, a.negbuckets = app2.posbuckets, app2.negbuckets
 	a.posbucketsDelta, a.negbucketsDelta = app2.posbucketsDelta, app2.negbucketsDelta
+
+	return hc, app2
```
Comment (beorn7) [inline tsdb/chunkenc/histo.go:395]: Here you return the newly created appender, but the caller (I guess) is supposed to just go on with the old appender (the receiver of the method), into which you have copied all the fields.

File: tsdb/head.go [hunk 2]
```diff
@@ -2624,6 +2624,24 @@ func (s *memSeries) appendHistogram(t int64, sh histogram.SparseHistogram, appen
...
+		// !ok -> we need to cut a new chunk
+		// ok but we have interjections -> existing chunk needs recoding before we can append our histogram
+		// ok and no interjections -> chunk is ready to support our histogram
+		if !ok {
+			c = s.cutNewHeadChunk(t, chunkenc.EncSHS, chunkDiskMapper)
+			chunkCreated = true
+
+		} else if len(posInterjections) > 0 || len(negInterjections) > 0 {
+			// new buckets have appeared. we need to recode all prior histograms within the chunk before we can process this one.
+			s.headChunk.chunk, s.app = app.Recode(posInterjections, negInterjections, sh.PositiveSpans, sh.NegativeSpans)
```
Comment (beorn7) [inline tsdb/head.go:2641]: Here you replace `s.app` with the returned appender. But I think the `Recode` method acts on the receiver appender already, and the returned appender is not the right one to use. WDYT?

File: tsdb/chunkenc/histo.go (hunk 1 above)
Comment (beorn7) [reply tsdb/chunkenc/histo.go:395]: So perhaps don't return app2 to avoid the confusion?

**Comment by beorn7** [comment]
I tried this in the demo setup.

Good news: time series don't disappear now.
Bad news: The numbers look really weird.

**Comment by beorn7** [comment]
Found a race:

==================
WARNING: DATA RACE
Write at 0x00c000f18387 by goroutine 19:
  github.com/prometheus/prometheus/tsdb/chunkenc.(*bstream).writeBit()
      /home/beorn/go/src/github.com/prometheus/prometheus/tsdb/chunkenc/bstream.go:75 +0x3cc
  github.com/prometheus/prometheus/tsdb/chunkenc.(*HistoAppender).writeSumDelta()
      /home/beorn/go/src/github.com/prometheus/prometheus/tsdb/chunkenc/histo.go:405 +0x2ae
  github.com/prometheus/prometheus/tsdb/chunkenc.(*HistoAppender).AppendHistogram()
      /home/beorn/go/src/github.com/prometheus/prometheus/tsdb/chunkenc/histo.go:326 +0xc44
  github.com/prometheus/prometheus/tsdb.(*memSeries).appendHistogram()
      /home/beorn/go/src/github.com/prometheus/prometheus/tsdb/head.go:2645 +0x155
  github.com/prometheus/prometheus/tsdb.(*headAppender).Commit()
      /home/beorn/go/src/github.com/prometheus/prometheus/tsdb/head.go:1639 +0x10bb
  github.com/prometheus/prometheus/tsdb.dbAppender.Commit()
      /home/beorn/go/src/github.com/prometheus/prometheus/tsdb/db.go:858 +0x42
  github.com/prometheus/prometheus/tsdb.(*dbAppender).Commit()
      <autogenerated>:1 +0x84
  github.com/prometheus/prometheus/storage.(*fanoutAppender).Commit()
      /home/beorn/go/src/github.com/prometheus/prometheus/storage/fanout.go:191 +0x75
  github.com/prometheus/prometheus/scrape.(*timeLimitAppender).Commit()
      <autogenerated>:1 +0x62
  github.com/prometheus/prometheus/scrape.(*scrapeLoop).scrapeAndReport.func1()
      /home/beorn/go/src/github.com/prometheus/prometheus/scrape/scrape.go:1196 +0x91
  github.com/prometheus/prometheus/scrape.(*scrapeLoop).scrapeAndReport()
      /home/beorn/go/src/github.com/prometheus/prometheus/scrape/scrape.go:1263 +0xfa1
  github.com/prometheus/prometheus/scrape.(*scrapeLoop).run()
      /home/beorn/go/src/github.com/prometheus/prometheus/scrape/scrape.go:1149 +0x44c

Previous read at 0x00c000f18387 by goroutine 26:
  [failed to restore the stack]

Goroutine 19 (running) created at:
  github.com/prometheus/prometheus/scrape.(*scrapePool).sync()
      /home/beorn/go/src/github.com/prometheus/prometheus/scrape/scrape.go:565 +0x1084
  github.com/prometheus/prometheus/scrape.(*scrapePool).Sync()
      /home/beorn/go/src/github.com/prometheus/prometheus/scrape/scrape.go:481 +0xa16
  github.com/prometheus/prometheus/scrape.(*Manager).reload.func1()
      /home/beorn/go/src/github.com/prometheus/prometheus/scrape/manager.go:197 +0x56

Goroutine 26 (running) created at:
  net/http.(*Server).Serve()
      /usr/local/go/src/net/http/server.go:3013 +0x644
  github.com/prometheus/exporter-toolkit/web.Serve()
      /home/beorn/go/pkg/mod/github.com/prometheus/exporter-toolkit@v0.5.1/web/tls_config.go:192 +0x2d1
  github.com/prometheus/prometheus/web.(*Handler).Run.func2()
      /home/beorn/go/src/github.com/prometheus/prometheus/web/web.go:594 +0xc4
==================

**Comment by beorn7** [comment]
Another race, sadly still not able to restore stack of goroutine 26 (the query path, I suppose):

```
==================
WARNING: DATA RACE
Write at 0x00c00026e2ea by goroutine 19:
  github.com/prometheus/prometheus/tsdb/chunkenc.(*bstream).writeBit()
      /home/beorn/go/src/github.com/prometheus/prometheus/tsdb/chunkenc/bstream.go:75 +0x5bb
  github.com/prometheus/prometheus/tsdb/chunkenc.(*bstream).writeBits()
      /home/beorn/go/src/github.com/prometheus/prometheus/tsdb/chunkenc/bstream.go:107 +0xc1
  github.com/prometheus/prometheus/tsdb/chunkenc.putInt64VBBucket()
      /home/beorn/go/src/github.com/prometheus/prometheus/tsdb/chunkenc/varbit_buckets.go:58 +0x1dc
  github.com/prometheus/prometheus/tsdb/chunkenc.(*HistoAppender).AppendHistogram()
      /home/beorn/go/src/github.com/prometheus/prometheus/tsdb/chunkenc/histo.go:323 +0xbe6
  github.com/prometheus/prometheus/tsdb.(*memSeries).appendHistogram()
      /home/beorn/go/src/github.com/prometheus/prometheus/tsdb/head.go:2645 +0x155
  github.com/prometheus/prometheus/tsdb.(*headAppender).Commit()
      /home/beorn/go/src/github.com/prometheus/prometheus/tsdb/head.go:1639 +0x10bb
  github.com/prometheus/prometheus/tsdb.dbAppender.Commit()
      /home/beorn/go/src/github.com/prometheus/prometheus/tsdb/db.go:858 +0x42
  github.com/prometheus/prometheus/tsdb.(*dbAppender).Commit()
      <autogenerated>:1 +0x84
  github.com/prometheus/prometheus/storage.(*fanoutAppender).Commit()
      /home/beorn/go/src/github.com/prometheus/prometheus/storage/fanout.go:191 +0x75
  github.com/prometheus/prometheus/scrape.(*timeLimitAppender).Commit()
      <autogenerated>:1 +0x62
  github.com/prometheus/prometheus/scrape.(*scrapeLoop).scrapeAndReport.func1()
      /home/beorn/go/src/github.com/prometheus/prometheus/scrape/scrape.go:1196 +0x91
  github.com/prometheus/prometheus/scrape.(*scrapeLoop).scrapeAndReport()
      /home/beorn/go/src/github.com/prometheus/prometheus/scrape/scrape.go:1263 +0xfa1
  github.com/prometheus/prometheus/scrape.(*scrapeLoop).run()
      /home/beorn/go/src/github.com/prometheus/prometheus/scrape/scrape.go:1149 +0x44c

Previous read at 0x00c00026e2ea by goroutine 26:
  [failed to restore the stack]

Goroutine 19 (running) created at:
  github.com/prometheus/prometheus/scrape.(*scrapePool).sync()
      /home/beorn/go/src/github.com/prometheus/prometheus/scrape/scrape.go:565 +0x1084
  github.com/prometheus/prometheus/scrape.(*scrapePool).Sync()
      /home/beorn/go/src/github.com/prometheus/prometheus/scrape/scrape.go:481 +0xa16
  github.com/prometheus/prometheus/scrape.(*Manager).reload.func1()
      /home/beorn/go/src/github.com/prometheus/prometheus/scrape/manager.go:197 +0x56

Goroutine 26 (running) created at:
  net/http.(*Server).Serve()
      /usr/local/go/src/net/http/server.go:3013 +0x644
  github.com/prometheus/exporter-toolkit/web.Serve()
      /home/beorn/go/pkg/mod/github.com/prometheus/exporter-toolkit@v0.5.1/web/tls_config.go:192 +0x2d1
  github.com/prometheus/prometheus/web.(*Handler).Run.func2()
      /home/beorn/go/src/github.com/prometheus/prometheus/web/web.go:594 +0xc4
==================
```

**Comment by beorn7** [comment]
About the weird numbers: It looks like whenever the span layout gets extended, the older numbers change drastically. Often they just revert to the same constant value. So I guess when iterating over the old chunk, something is off, either with reading the old data or inserting it into the new chunk.

File: tsdb/chunkenc/histo.go (hunk 1 above)
Comment (Dieterbe) [reply tsdb/chunkenc/histo.go]: no the caller updates the chunk and the appender in head.go, with this code:
```
		} else if len(posInterjections) > 0 || len(negInterjections) > 0 {
			// new buckets have appeared. we need to recode all prior histograms within the chunk before we can process this one.
			s.headChunk.chunk, s.app = app.Recode(posInterjections, negInterjections, sh.PositiveSpans, sh.NegativeSpans)
		}
```

this actually makes it pointless to copy all those fields back into the old appender, so i took that out.

**Comment by beorn7** [comment]
Will merge now, tests to be fixed in separate PRs.

**Pull request merged**